File: backend/models/Task.js
const mongoose = require("mongoose");

const TaskSchema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: [true, "タスク名を入力してください"],
    },
    description: String,
    startDateAt: Date,
    endDateAt: Date,
    state: {
      type: Number,
      enum: [0, 1, 2],
    },
    progress: {
      type: Number,
      min: 0,
      max: 100,
    },
    priority: {
      type: Number,
      enum: [0, 1, 2],
    },
    todaysTask: Boolean,
    project: { type: mongoose.Schema.ObjectId, null: true, ref: "Project" },
    workspace: {
      type: mongoose.Schema.ObjectId,
      required: true,
    },
    user: {
      type: mongoose.Schema.ObjectId,
      ref: "User",
      required: true,
    },
  },
  {
    timestamps: true,
  }
);

module.exports = mongoose.model("Task", TaskSchema);

File: backend/app/models/task.rb
# == Schema Information
#
# Table name: tasks
#
#  id            :bigint           not null, primary key
#  description   :string(255)
#  end_date_at   :datetime
#  name          :string(255)      not null
#  priority      :integer          default("low")
#  progress      :integer          default(0)
#  start_date_at :datetime
#  state         :integer          default("open")
#  todays_task   :boolean          default(FALSE)
#  created_at    :datetime         not null
#  updated_at    :datetime         not null
#  project_id    :bigint
#  user_id       :bigint           not null
#  workspace_id  :bigint           not null
#
# Indexes
#
#  index_tasks_on_user_id_and_project_id    (user_id,project_id)
#  index_tasks_on_user_id_and_workspace_id  (user_id,workspace_id)
#
class Task < ApplicationRecord
  belongs_to :user
  belongs_to :workspace
  belongs_to :project

  validates :name, :presence => true
  validates :progress, :numericality => {
    :greater_than_or_equal_to => 0,
    :less_than_or_equal_to => 100
  }

  enum :state => { :open => 0, :progress => 1, :done => 2 }
  enum :priority => { :low => 0, :normal => 1, :high => 2 }
end
